crates/node/src/roles/worker.rs

```diff
@@ -86,73 +85,140 @@ impl WorkerRole {
-        ));
+        let mut component_set = JoinSet::new();
 
         // Ensures bifrost has initial metadata synced up before starting the worker.
-        let mut bifrost_handle = self.bifrost.start(inner_shutdown_watch.clone()).await?;
+        let mut bifrost_join_handle = self.bifrost.start(inner_shutdown_watch.clone()).await?;
 
-        let mut worker_handle = tokio::spawn(self.worker.run(inner_shutdown_watch));
+        component_set.spawn(
+            self.worker
+                .run(inner_shutdown_watch)
+                .map_ok(|_| "worker")
+                .map_err(WorkerRoleError::Worker),
+        );
+        component_set.spawn(
+            Self::reload_schemas(
+                subscription_controller,
+                self.schemas,
+                // todo: make this configurable
+                Channel::builder("http://127.0.0.1:5122/".parse().expect("valid uri"))
+                    .connect_lazy(),
+            )
+            .map_ok(|_| "schema-update"),
+        );
 
         tokio::select! {
             _ = shutdown_signal => {
                 info!("Stopping worker role");
-                let _ = tokio::join!(inner_shutdown_signal.drain(), admin_handle, meta_handle, worker_handle, bifrost_handle);
-            },
-            result = &mut meta_handle => {
-                result.map_err(WorkerRoleError::MetaPanic)??;
-                panic!("Unexpected termination of meta.");
+                inner_shutdown_signal.drain().await;
+                // ignoring result because we are shutting down
+                let _ = tokio::join!(component_set.shutdown(), bifrost_join_handle);
             },
-            result = &mut admin_handle => {
-                result.map_err(WorkerRoleError::AdminPanic)??;
-                panic!("Unexpected termination of admin.");
-            },
-            result = &mut worker_handle => {
-                result.map_err(WorkerRoleError::WorkerPanic)??;
-                panic!("Unexpected termination of worker.");
+            Some(component_result) = component_set.join_next() => {
+                let component_name = component_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of component '{component_name}'");
+            }
+            bifrost_result = &mut bifrost_join_handle => {
+                bifrost_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of bifrost service");
             }
-            result = &mut bifrost_handle => {
-                result.map_err(WorkerRoleError::BifrostPanic)??;
-                panic!("Unexpected termination of bifrost service.");
-            },
         }
 
         Ok(())
     }
+
+    async fn reload_schemas<SC>(
+        subscription_controller: Option<SC>,
+        schemas: Schemas,
+        schema_channel: Channel,
+    ) -> Result<(), WorkerRoleError>
+    where
+        SC: SubscriptionController + Clone + Send + Sync,
+    {
```

Comment:
This, among other bits, are candidates for a new WatchDog component(s) that I was planning to build. Let's discuss the details offline.